### ISSUE TITLE
Lightcurves from the new "TICA" products?

### ISSUE THREAD

**Issue by gully** [lightkurve/lightkurve#993]
[TICA data products](https://archive.stsci.edu/hlsp/tica) are now provided ahead of other official TESS data releases under a pilot program that started yesterday.  How could these products be read with lightkurve?  We used to have the TPFfactory, is that still around and amenable to adaptation for these TICA products?

@mrtommyb and others at TESS GI may have thoughts?

**Comment by barentsen** [comment]
Thanks @gully. This is an excellent question, worth discussing!

There still is a TPFfactory in Lightkurve, but it needs some review, testing, and documentation to verify that it works with TICA products.  A disadvantage of the tool is that it requires FFIs to be available locally, so it's not as seamless as using TessCut.

An alternative approach would be to have an API like TessCut be available for TICA FFIs, but that is not cheap to set up and maintain.

I wonder if what we need is a client-side application that enables cutouts to be made from online or cloud-hosted FFIs using the "range request" mechanism available in HTTP and S3 to retrieve specific bytes from large files.  I started prototyping this idea in two packages, both of which are still ~vaporware:

https://github.com/SSDataLab/tess-bite
https://github.com/SSDataLab/tess-cloud

**Comment by gully** [comment]
Whoa far out ideas here!  Glad you've thought about this.  I had experimented with making TPFfactory a little more generalized, there may be an un-merged Pull Request somewhere.  I got it to read Spitzer frames for making IRAC lightcurves, rather than K2 superstamps.  The local files part is tricky since the FFIs are so big, but maybe self-limiting?  Folks will only download all the FFIs if they really need an answer quickly, otherwise can wait for the standard products?

**Comment by mrtommyb** [comment]
Right now we are only trialing the release of these products, although I think it's fair to say that the long-term plan at TESS is for rapid data release. You are right that currently these data are not linked to other things at the MAST and do not interface with TESSCut. 

I suspect that designing anything around the current TICA products might be a little premature because there may a future where the releases are more closely integrated within the MAST ecosystem (I don't have any additional information, just guessing).

**Comment by scfleming** [comment]
tl;dr, I wouldn't hold your breath for these to end up in TESSCut.

It's up to the lightkurve team of course how and when they want to add support for TICA in lightkurve given that this is a beta, with only the promise of Sector 35 being released early, but there is nothing in our current roadmap at MAST to hook these up into TESSCut any time soon, unless there is a specific direction to do so and accompanying resources from the mission level to make that happen (this is NOT cheap for us to do, and is not scoped in the current mission contract for MAST).  Put another way, if we were in a race to see if lightkurve or MAST would get a cutout functionality working with TICA products, I'd put my money on lightkurve at this point.

**Comment by barentsen** [comment]
I went ahead and added a prototype version of TessCut for TICA to my experimental `tess-cloud` package.  It runs entirely on the client side so no API is needed.  @gully Can you try this out?

Syntax:

```python
from tess_cloud import tica
images = tica.list_tica_images(sector=35, orbit=1, camera=1, ccd=1)
tpf = images.cutout(col=103, row=121, shape=(5, 5))
```

I was able to cut a 5x5 TPF from 1712 TICA images over wifi in 12 seconds (because I download the FFI data in a sick way using HTTP-range requests and `asyncio`). Video demo:

https://user-images.githubusercontent.com/817669/110895672-09309a80-82af-11eb-9937-50d2b60219f5.mp4

**Comment by gully** [comment]
This is awesome!! I will be happy to try this out, but at first sight it looks as advertised! I'm amazed range requests works and how you reverse engineered which-bytes-are-where, so maybe I'll take a peak under the hood.

So this advance is really useful, it means that with TICA products released promptly, plus this tool, users can get to a lightcurve much faster than before, possibly after the first orbit of a sector has telemetered data, but before the second orbit of a sector has completed (depending on how fast the TICA products are made available).  So I would encourage the TESS team to continue offering the TICA products on this advanced schedule.  Bravo!

**Comment by mrtommyb** [comment]
The Sector 35, Orbit 2 data is online. I'm interested in seeing some real-world examples of the data being used, but maybe we need a tool like this one to see great uptake in the use of these data.

**Comment by gully** [comment]
Hi @barentsen I am trying to install `tess-cloud` and get hung up at the pip install phase, please advise.

**Comment by barentsen** [comment]
I'm planning to make a proper package out of this today. Stay tuned...

**Comment by christopherburke** [comment]
Oh wow Geert this looks really awesome. I learned something today about http range partial downloads. Ive never understood who is the one paying for all the downloading of TESS data on S3, does STScI pay because they have the data 'bucket' accessible to public? Presumably this could in principle be done against the regular MAST storage area, but in a non-asynchronus manner.

**Comment by barentsen** [comment]
@christopherburke The TICA FFIs aren't actually on AWS S3 yet, so I've modified `tess-cloud` to use range requests from the MAST HTTP server instead.

I'll make the cutout tool available in a proper package this weekend so you can advertise it to all TICA fans!

I can't answer who pays to make the data accessible, but it's definitely true that novel science tools can only be developed if the mission/archive/nasa continue to serve the data in flexible and free way like this.  I made folks at MAST aware that this tool hits MAST with a significant number of requests, and I believe their guidance is to keep going until it appears to become an issue. Fingers crossed we can keep doing this!

**Comment by scfleming** [comment]
To echo, when you do advertise please let us know.  We'll monitor traffic and if there are too many requests and impact servers we might need to limit rates or ask you to turn it off.

**Comment by gully** [comment]
Hey all, I attemped the `from_fits_images` way of making a TPF from the local TICA FFIs and encountered a bug:

```python
from lightkurve import TessTargetPixelFile
fnames = np.sort(tica_fns)
tpf = TessTargetPixelFile.from_fits_images(images_flux=fnames,
                                             position=SkyCoord(lc.RA, lc.Dec, unit='deg'),
                                             extension=0,
                                             size=(34,34))
```

`KeyError: "Keyword 'CRVAL1P' not found."`

Has anyone successfully run `from_fits_images` on TESS TICA?

**Comment by gully** [comment]
In regards to tess-cloud, `pip install .` started to work, but threw an error about conflicts.  I got it to work after messing around with poetry, but hit an AWS credential issue (I just need to configure AWS for boto3 I think).

```
ERROR: Could not find a version that satisfies the requirement tess-locator>=0.2.1 (from tess-cloud)
ERROR: No matching distribution found for tess-locator>=0.2.1
```

**Comment by gully** [comment]
So recapping here in regards to @mrtommyb 's statement
>  I'm interested in seeing some real-world examples of the data being used, but maybe we need a tool like this one to see great uptake in the use of these data.

`TessTargetPixelFile.from_fits_images` did not work out of the box, but I think---with some spot-checking of FITS headers---it  can work.  @mrtommyb  is there someone in the TESS GI Office who could make a patch to lightkurve?

tess-cloud is under active development and is using some technology that is so futuristic that several of us here are mouths agape. :rocket: :robot: :trophy: 

But that leaves no plug-and-play solution for TICA at the moment.  I ultimately made a custom TICA lightcurve using quicklook numpy array manipulation.  Here's a peek:

![custom_TICA_sector35_orbit1](https://user-images.githubusercontent.com/860227/111833705-8b791a00-88c0-11eb-842d-d0a85ccbee7f.png)

The vertical lines show contemporaneous ancillary ground-based measurements.  Having Early Release TICA is useful since the realized phases can inform the extent to which we can cease ground-based acquisitions, add additional observations, or apply constraints to the timing of future observations, all of which act to make telescope allocation more efficient and effective, saving our scarce resources.  This one orange lightcurve took 30 GB of FFIs, so repeating this for several sources on different CCDs is a bit unwieldy, but possible for those who really need it.  So while there is a cost to patching `TessTargetPixelFile.from_fits_images` and serving the massive dataset, there is also a real dollar value to making it possible, and doubly so for making it fast and easy.

**Comment by rebekah9969** [comment]
Hello @gully 

With regards to your last point about one trick you could try is the following, 

```
hdul = fits.open(fname)
hdr = hdul[0].header 
hdr.append(('CRVAL1P', 0.0), end=True)
hdr.append(('CRVAL2P', 0.0), end=True)
hdul.close()
```

This should add the variables you need to get it to act like a tpf. This is only a work around however and I will work on something more concrete.

Thanks

**Comment by scfleming** [comment]
We are starting to get concerned with the number of connections used for TICA cutouts here.  Is there a way you can reduce the number of connections per second?  We can also explore setting limits here for all requests asking for "TICA" products.

**Comment by barentsen** [comment]
Uh oh!  That's concerning because this would only be traffic generated while testing/developing the feature.

Being able to fire a large number of requests simultaneously in an asynchronous way is key to making the tool performant, otherwise it just ends up waiting for i/o all the time.

I optimized the tool last week so that it only requires 1 request per row of data, so that cutting out a 5x5 rectangle from 2000 images triggers 10,000 requests.  That's of course a large number still.

The data transferred by each request is very low (<kb).  Is the concern that 10,000 requests per cutout create a significant CPU load?

**Comment by scfleming** [comment]
There's too many connections that are opened from one IP address.  Everything flows through one machine that services archive.stsci.edu, so if one or a few IP addresses doing this may take up ALL available connections on the archive.stsci.edu server, impacting others who are looking for other files from other missions, or even webpages.

**Comment by scfleming** [comment]
How long would it take if there were 1,000 vs. 100 requests compared to the 10,000 being used now?

**Comment by barentsen** [comment]
I guess there is an important difference between the number of connections and the number of requests.  Thank you for emphasizing this!

My best understanding is that `tess-cloud` limits itself to a pool of 100 simultaneous connections (because this appears to be the default for `aiohttp.BaseConnector`) and does not fire more than 200 requests in parallel (by virtue of the `MAX_CONCURRENT_DOWNLOADS` semaphore within tess-cloud).

Would you recommend lowering these settings?

I will go ahead and profile the connection behavior in more detail to verify the behavior and identify any bugs or optimizations.

**Comment by barentsen** [comment]
Update: it looks like someone may be hammering MAST with tess-cloud right now.

I am going to make the tess-cloud repository private and yank the releases from pypi for now until I can thoroughly analyze and improve the connection pooling behavior, 

If anyone who reads this is actively using tess-cloud, please stop until further notice!

**Comment by barentsen** [comment]
I set up a mock MAST TICA file server ([details here](https://github.com/SSDataLab/tess-cloud/tree/main/mockserver)) which allowed me to verify the connection behavior of `tess-cloud`.

My findings:
* I confirmed that the number of TCP connections used by `tess-cloud` is controlled by the `limit` parameter of the `aiohttp.TCPConnector` class, which defaults to `limit=100`.
* I decreased this default from 100 to 3 to be more kind to the server.  This is determined by a new constant: `tess_cloud.MAX_TCP_CONNECTIONS = 3`.
* As a result, TICA Sector 36 cutouts now take between 1 and 10 minutes, depending on connection latency (cloud vs wifi), but it hopefully makes the tool more sustainable.
* All requests will now be made using a unique user agent identifier to make it easier for MAST to identify traffic from this tool and throttle or ban problematic versions.

I uploaded tess-cloud v0.3.1 to pypi to include these changes.  Please do not use earlier versions!  Traffic from this version will identify itself to MAST as `User-Agent: tess-cloud/v0.3.1`.

The load could further be reduced by requesting bigger chunks, i.e. by making fewer requests at the cost of using extra bandwidth.  I will investigate this route if the load continues to be unsustainable.
